nuxeo-core/nuxeo-core/src/main/java/org/nuxeo/ecm/core/trash/AbstractTrashService.java

```diff
@@ -112,19 +117,60 @@ public boolean canPurgeOrUntrash(List<DocumentModel> docs, NuxeoPrincipal princi
             return false;
         }
         // used to do only check on parent perm
-        TrashInfo info = getInfo(docs, principal, false, true);
+        TrashInfo info = getInfo(docs, principal, false, true, true);
         return info.docs.size() == docs.size();
     }
 
     protected TrashInfo getInfo(List<DocumentModel> docs, NuxeoPrincipal principal, boolean checkProxies,
             boolean checkDeleted) {
+        return getInfo(docs, principal, checkProxies, checkDeleted, false);
+    }
+
+    @Override
+    public boolean getCanEmptyTrash(DocumentModel parent, NuxeoPrincipal principal) {
+        String nxql = String.format(TRASHED_QUERY, parent.getId());
+        ScrollRequest request = DocumentScrollRequest.builder(nxql)
+                                                     .username(principal.getName())
+                                                     .repository(parent.getRepositoryName())
+                                                     .build();
+        ScrollService service = Framework.getService(ScrollService.class);
+        boolean canPurge = true;
+        try (Scroll scroll = service.scroll(request)) {
+            if (!scroll.hasNext()) {
+                return false;
+            }
+            while (scroll.hasNext() && canPurge) {
+                DocumentModelList documents = RepositoryScroll.loadDocuments(parent.getCoreSession(), scroll.next());
+                canPurge = canPurgeOrUntrash(documents, principal);
+            }
+        }
+        return canPurge;
+    }
+
+    /**
+     * Return the {@link TrashInfo} of a document list with potentially some documents not ok for 'purge' or unTrash
+     * (forbidden, locked ...)
+     *
+     * @param docs the documents List to trash
+     * @param principal the user principal
+     * @param checkProxies <code>true</code> to check proxies.
+     * @param checkDeleted <code>true</code> to check if doc is trashed.
+     * @param exitOnFirstDelta <code>true</code> to avoid looping in the whole documents List and exist on first
+     *            document with TrashInfo "forbidden", "proxies" or "locked".
+     * @return the TrashInfo for the documents List.
+     */
+    protected TrashInfo getInfo(List<DocumentModel> docs, NuxeoPrincipal principal, boolean checkProxies,
+            boolean checkDeleted, boolean exitOnFirstDelta) {
         TrashInfo info = new TrashInfo();
         info.docs = new ArrayList<>(docs.size());
         if (docs.isEmpty()) {
             return info;
         }
         CoreSession session = docs.get(0).getCoreSession();
         for (DocumentModel doc : docs) {
+            if (exitOnFirstDelta && (info.forbidden > 0 || info.locked >0 || info.proxies >0)){
+                return info;
```

Comment:
With this premature exit you change/break the exiting behaviour of this method as before you changed it, it main purpose was to give more informations about the trash state, and now you exit the method if your check is true  i think we should rework it depending on the need

Reply:
The behavior is only changed if the exitOnFirstDelta is true, which by default is false in the new internal call so the behavior is still the same:

https://github.com/nuxeo/nuxeo/pull/3693/commits/d61fa8ff5b9e46aa598002900f0e794c635bc76b#diff-7c7663426b7649bd97453eaa63571db0R122

Not advocating for a way or another though

Reply:
finally roll-backed to reduice commit impact to jsf bean only

Reply:
@NourNuxeo about https://github.com/nuxeo/nuxeo/pull/3693#discussion_r372988424 here the method is changed on its behaviour, before adding parameters the method will take each parameter as a filter for the process but the new once will exit the loop which is different.

as discussed with thierry this change is not required and it's why he rollback it

Reply:
Of course it does extra stuff, that's why he changes it. But it still behaves the same way in the places it was called before.

Yes I followed up with the discussion about the rollback 👍

Reply:
> Of course it does extra stuff, that's why he changes it. But it still behaves the same way in the places it was called before.

i know why he changed it,  but i don't agree with this modification as the new method now can do more then just filter and any one can called it without knowing exactly what means the new parameter

Reply:
The method was protected and the param was named. But I like method who do only one thing better than the multi behavioral ones for sure.